ssz_serialization/codec.nim

```diff
@@ -97,155 +98,35 @@ template checkForForbiddenBits(ResulType: type,
-    if recList.len == 0 and enumVal != 0:
-      macros.error("only None discriminator can have empty case branch", recList)
-
-  result = newEmptyNode()
-
-macro unionSizeImpl(id: typed, x: type): untyped =
-  let
-    T = x.getType[1]
-    recList = T.getTypeImpl[2]
-    recCase = recList[0]
-    kind = recCase[0][0]
-    TKind = recCase[0][1]
-
-  var hasNull = false
-
-  result = quote do:
-    case `id`.`kind`
-    of `TKind`(0): 1
-
-  # begin with 1: skip the discriminator
-  for i in 1..<recCase.len:
-    let branch = recCase[i]
-    if branch.kind == nnkOfBranch:
-      let recList = branch[1]
-      let enumVal = branch[0].intVal
-      if enumVal == 0:
-        hasNull = true
-      else:
-        let field = recList[0][0]
-        result.add nnkOfBranch.newTree(
-          (quote do: `TKind`(`enumVal`)),
-          quote do: 1 + sszSize(`id`.`field`)
-        )
+macro doInit[T](
+    t: typedesc[T], selectorKey: static string, selector: typed): T =
+  let selectorId = ident(selectorKey)
+  quote do:
+    when compiles(`t`(`selectorId`: `selector`)):
+      `t`(`selectorId`: `selector`)
     else:
-      # else branch
-      let recList = branch[0]
-      let field = recList[0][0]
-      result.add nnkElse.newTree(
-        quote do: 1 + sszSize(`id`.`field`)
-      )
-
-  if not hasNull:
-    macros.error("no null branch detected", recCase)
-
-func unionSize*(val: auto): int {.gcsafe, raises:[].} =
-  type T = type val
-  unionSizeImpl(val, T)
-
-macro initSszUnionImpl(RecordType: type, input: openArray[byte]): untyped =
-  var res = newStmtList()
-  let TInst = RecordType.getTypeInst[1]
-  let T = RecordType.getType[1]
-  var recordDef = getImpl(T)
-
-  for field in recordFields(recordDef):
-    if (field.isDiscriminator):
-      let
-        selectorFieldName = field.name
-        selectorFieldType = field.typ
-        SelectorFieldNameLit = newLit($selectorFieldName)
-
-      res.add quote do:
-        block:
-          if `input`.len == 0:
-            raiseMalformedSszError(`type recordDef`, "empty union not allowed")
-
-          var selector: `selectorFieldType`
-          # `checkedEnumAssign` also checks for holes in an enum.
-          if not checkedEnumAssign(selector, `input`[0]):
-            raiseMalformedSszError(`type recordDef`, "union selector is out of bounds")
-
-          var caseObj = `TInst`(`selectorFieldName`: selector)
-
-          var fieldCount = 0
-          enumInstanceSerializedFields(caseObj, fieldName, field):
-            when fieldName != `SelectorFieldNameLit`:
-              readSszBytes(`input`.toOpenArray(1, `input`.len - 1), field)
-              fieldCount.inc()
-
-          if fieldCount == 0: # This represents a `None` in the Union
-            if `input`.len != 1:
-              raiseMalformedSszError(`type recordDef`, "Union None should have no value")
-
-          caseObj
-
-      break
-
-  res
+      `t`.init(`selectorId` = `selector`)
```

Comment:
This is the synthesized `init` from https://github.com/status-im/nim-serialization/pull/105 for unions with holey selector enum, regular unions take the `when compiles()` path.